src/coreclr/jit/morph.cpp

```diff
@@ -13336,97 +13336,97 @@ GenTree* Compiler::fgOptimizeRelationalComparisonWithCasts(GenTreeOp* cmp)
+            assert(varTypeIsLong(op->CastToType()));
+            return true;
+        }
+
+        return false;
+    };
+
+    if (!supportedOp(op1) || !supportedOp(op2))
     {
-        // op1 is either a LONG constant (yes, it's pretty normal for relops)
-        // or (T)ARR_LENGTH
-        castOp             = cmp->gtGetOp2();
-        knownPositiveOp    = cmp->gtGetOp1();
-        knownPositiveIsOp2 = false;
+        return cmp;
     }
 
-    if (castOp->OperIs(GT_CAST) && varTypeIsLong(castOp->CastToType()) && castOp->AsCast()->CastOp()->TypeIs(TYP_INT) &&
-        castOp->IsUnsigned() && !castOp->gtOverflow())
-    {
-        bool knownPositiveFitsIntoU32 = false;
-        if (knownPositiveOp->IsIntegralConst() && FitsIn<UINT32>(knownPositiveOp->AsIntConCommon()->IntegralValue()))
-        {
-            // BTW, we can fold the whole condition if op2 doesn't fit into UINT_MAX.
-            knownPositiveFitsIntoU32 = true;
-        }
-        else if (knownPositiveOp->OperIs(GT_CAST) && varTypeIsLong(knownPositiveOp->CastToType()) &&
-                 knownPositiveOp->AsCast()->CastOp()->OperIs(GT_ARR_LENGTH))
+    auto isUpperZero = [this](GenTree* op) {
+        if (op->IsIntegralConst())
         {
-            knownPositiveFitsIntoU32 = true;
-            // TODO-Casts: recognize Span.Length here as well.
+            int64_t lng = op->AsIntConCommon()->LngValue();
+            return (lng >= 0) && (lng <= UINT_MAX);
         }
 
-        if (!knownPositiveFitsIntoU32)
+        assert(op->OperIs(GT_CAST));
+        if (op->AsCast()->IsUnsigned())
         {
-            return cmp;
+            return true;
         }
 
+        return IntegralRange(SymbolicIntegerValue::Zero, SymbolicIntegerValue::IntMax)
+            .Contains(IntegralRange::ForNode(op->AsCast()->CastOp(), this));
+    };
+
+    bool op1UpperIsZero = isUpperZero(op1);
+    bool op2UpperIsZero = isUpperZero(op2);
+
+    // If both operands have zero as the upper half we can optimize then any
+    // signed/unsigned 64-bit comparison is equivalent to the same unsigned
+    // 32-bit comparison.
+    if (op1UpperIsZero && op2UpperIsZero)
+    {
         JITDUMP("Removing redundant cast(s) for:\n")
         DISPTREE(cmp)
         JITDUMP("\n\nto:\n\n")
 
         cmp->SetUnsigned();
 
-        // Drop cast from castOp
-        if (knownPositiveIsOp2)
-        {
-            cmp->gtOp1 = castOp->AsCast()->CastOp();
-        }
-        else
-        {
-            cmp->gtOp2 = castOp->AsCast()->CastOp();
-        }
-        DEBUG_DESTROY_NODE(castOp);
-
-        if (knownPositiveOp->OperIs(GT_CAST))
-        {
-            // Drop cast from knownPositiveOp too
-            if (knownPositiveIsOp2)
+        auto transform = [this](GenTree** use) {
+            if ((*use)->IsIntegralConst())
             {
-                cmp->gtOp2 = knownPositiveOp->AsCast()->CastOp();
+                (*use)->BashToConst(static_cast<int>((*use)->AsIntConCommon()->LngValue()), TYP_INT);
```

Comment:
Nit: no need to specify `TYP_INT`, the idea with `BashToConst` is that `static_cast<int>` implies it.